DataFormats/Headers/include/Headers/DataHeader.h

```diff
@@ -485,6 +485,28 @@ auto get(const void* buffer, size_t len = 0)
   return get<HeaderType>(reinterpret_cast<const byte *>(buffer), len);
 }
 
+template <typename T>
+T strtoui(const char *str, char **str_end, int base ) {
+  if constexpr (sizeof(T)==sizeof(uint32_t)) {
+    T res = std::strtoul(str, str_end, base);
+    if (res<=std::numeric_limits<uint32_t>::max()) {
```

Comment:
always true?
why not `return std::stoul()` ?

Reply:
that's a bit of paranoia: unsigned long (which strtoul returns) is *at least* 32bit long, whereas uint32_t is *exactly* 32bits long. Where unsigned int is the same as uint32_t the compiler does the correct thing and optimizes all of that away. But there is indeed a bug: res should be unsigned int :) and maybe the strtoul error conditions should be handled - will do.

Reply:
after a quick check it seems unsigned long is 64bit (with gcc on x86 at least), but the new code works as expected: https://godbolt.org/z/wJgaid

Reply:
length of long is compiler dependent... Stupid idea in the c standard not to fix the size of integers.
I'd use int for 32 or long long for 64, or just uint32_t / uint64_t.

Reply:
yes, LP64 has long at 64. Sould be fine.
The code you link to fails to test the limits for 64bit types. That test is always true.

Reply:
@ironMann it does not need to test the range if the type is as long as unsigned long long - it gets optimized away. you can verify by changing the test slightly to do e.g. "<" instead of "<="

Reply:
...and the size of unsigned long long is implementation dependent, so the check needs to be there... but it all happens at compile time so no necessary overhead :)

Reply:
long long is at least 64bit. and it is exactly 64bit in the LP64 model. The code as is will never throw if you return uint64_t. Why not test `errno` instead?

Reply:
it will never throw as long as long long is 64bits, but it might not be somewhere else, could be 128. I should probably check errno anyway, but also the range check needs to be there.

Reply:
errno checking added, ruining the nice optimizations :)

Reply:
complaining about checking one value after converting a string to integer :)

seriously, i think the test after stroul(l) should be:
```cpp
unsigned long long res = std::strtoull(str, str_end, base);
if (errno == ERANGE || *str_end != (str + std::strlen(str))) {
    throw std::range_error("value too big or invalid");
} else {
    return static_cast<T>(res);
}
```
1. errno covers overflows
2. tests the validity of input string. We want all characters interpreted as digits.

Reply:
@ironMann indeed I agree we can require the string to be a pure number and pass the validity checks, i'll add the check. It looks like you're still suggesting to drop the range check: long long could still be larger than 64bit producing a value that is too large for the target type, so i'm leaving that check in.

Reply:
@ ironMann, actually i would not go beyond the strtoul api since we are emulating that interface here - so i'd leave out the check for full consumption: it is up to the user to check str_end. So only on range error we crap out with an exception, otherwise we just return whatever strtoull returns.

Reply:
so after some thinking I changed the implementation to simply wrap strtoul with only the additional check for range, also made it noexcept so it supplements it's std library counterpart.
Also added the "C++" equivalent: stoui - also here it is just a "strongly typed" wrapper for std::stoul with the additional range check. this one throws.